theseus/src/api/profile_create.rs

```diff
@@ -0,0 +1,154 @@
+//! Theseus profile management interface
+use crate::{prelude::ModLoader, profile};
+pub use crate::{
+    state::{JavaSettings, Profile},
+    State,
+};
+use daedalus::modded::LoaderVersion;
+use dunce::canonicalize;
+use futures::prelude::*;
+use std::path::PathBuf;
+use tokio::fs;
+use tokio_stream::wrappers::ReadDirStream;
+use uuid::Uuid;
+
+const DEFAULT_NAME: &str = "Untitled Instance";
+
+// Generic basic profile creation tool.
+// Creates an essentially empty dummy profile with profile_create
+#[tracing::instrument]
+pub async fn profile_create_empty() -> crate::Result<PathBuf> {
+    profile_create(
+        String::from(DEFAULT_NAME), // the name/path of the profile
+        String::from("1.8.2"),      // the game version of the profile
+        ModLoader::Vanilla,         // the modloader to use
+        String::from("stable"), // the modloader version to use, set to "latest", "stable", or the ID of your chosen loader
+        None,                   // the icon for the profile
+    )
+    .await
+}
+
+// Creates a profile at  the given filepath and adds it to the in-memory state
+// Returns filepath at which it can be accessed in the State
+#[tracing::instrument]
+pub async fn profile_create(
+    name: String,           // the name of the profile, and relative path
+    game_version: String,   // the game version of the profile
+    modloader: ModLoader,   // the modloader to use
+    loader_version: String, // the modloader version to use, set to "latest", "stable", or the ID of your chosen loader
+    icon: Option<PathBuf>,  // the icon for the profile
+) -> crate::Result<PathBuf> {
+    let state = State::get().await?;
+
+    let uuid = Uuid::new_v4();
+    let path = state.directories.profiles_dir().join(uuid.to_string());
+
+    if path.exists() {
+        if !path.is_dir() {
+            return Err(ProfileCreationError::NotFolder.into());
+        }
+        if path.join("profile.json").exists() {
+            return Err(ProfileCreationError::ProfileExistsError(
+                path.join("profile.json"),
+            )
+            .into());
+        }
+
+        if ReadDirStream::new(fs::read_dir(&path).await?)
+            .next()
+            .await
+            .is_some()
+        {
+            return Err(ProfileCreationError::NotEmptyFolder.into());
+        }
+    } else {
+        fs::create_dir_all(&path).await?;
+    }
+    println!(
+        "Creating profile at path {}",
+        &canonicalize(&path)?.display()
+    );
+
+    let loader = modloader;
+    let loader = if loader != ModLoader::Vanilla {
+        let version = loader_version;
+
+        let filter = |it: &LoaderVersion| match version.as_str() {
+            "latest" => true,
+            "stable" => it.stable,
+            id => it.id == *id,
+        };
+
```

Comment:
forge in a lot of cases doesn't have a stable version of the mod loader set. in those cases I think it's better to fallback to latest instead of failing. could that functionality be added here?